File: adapter/src/main/java/com/github/mrmitew/bodylog/adapter/common/presenter/DetachableMviPresenter.java
package com.github.mrmitew.bodylog.adapter.common.presenter;

import com.github.mrmitew.bodylog.adapter.common.view.BaseView;

public abstract class DetachableMviPresenter<V extends BaseView<S>, S> extends BaseMviPresenter<V, S> implements HasDetachableView<V> {
    protected DetachableMviPresenter(final V view) {
        super(view);
    }

    @Override
    public void attachView(final V view) {
        mView = view;
    }

    @Override
    public void detachView() {
        mView = null;
        unbindIntents();
    }
}

File: adapter/src/main/java/com/github/mrmitew/bodylog/adapter/common/presenter/DetachableMviPresenter.kt
package com.github.mrmitew.bodylog.adapter.common.presenter

import com.github.mrmitew.bodylog.adapter.common.UiState
import com.github.mrmitew.bodylog.adapter.common.view.BaseView

abstract class DetachableMviPresenter<V : BaseView<S>, S : UiState> protected constructor(view: V) :
        BaseMviPresenter<V, S>(view), HasDetachableView<V> {

    override fun attachView(view: V) {
        this.view = view
    }

    override fun detachView() {
        this.view = null
        unbindIntents()
    }
}
